ddtrace/debugging/_exception/auto_instrument.py

```diff
@@ -0,0 +1,204 @@
+)
+
+
+def unwind_exception_chain(
+    exc,  # type: t.Optional[BaseException]
+    tb,  # type: t.Optional[TracebackType]
+):
+    # type: (...) -> t.Tuple[t.Deque[t.Tuple[BaseException, t.Optional[TracebackType]]], t.Optional[uuid.UUID]]
+    """Unwind the exception chain and assign it an ID."""
+    chain = deque()  # type: t.Deque[t.Tuple[BaseException, t.Optional[TracebackType]]]
+
+    while exc is not None:
+        chain.append((exc, tb))
+
+        try:
+            if exc.__cause__ is not None:
+                exc = exc.__cause__
+            elif exc.__context__ is not None and not exc.__suppress_context__:
+                exc = exc.__context__
+            else:
+                exc = None
+        except AttributeError:
+            # Python 2 doesn't have exception chaining
+            break
+
+        tb = getattr(exc, "__traceback__", None)
+
+    exc_id = None
+    if chain:
+        # If the chain is not trivial we generate an ID for the whole chain and
+        # store it on the outermost exception, if not already generated.
+        exc, _ = chain[-1]
+        try:
+            exc_id = exc._dd_exc_id  # type: ignore[attr-defined]
+        except AttributeError:
+            exc._dd_exc_id = exc_id = uuid.uuid4()  # type: ignore[attr-defined]
+
+    return chain, exc_id
+
+
+@attr.s
+class SpanExceptionProbe(LogLineProbe):
+    @classmethod
+    def build(cls, exc_id, tb):
+        # type: (uuid.UUID, TracebackType) -> SpanExceptionProbe
+        _exc_id = str(exc_id)
+        frame = tb.tb_frame
+        filename = frame.f_code.co_filename
+        line = tb.tb_lineno
+        name = frame.f_code.co_name
+        message = "exception info for %s, in %s, line %d (exception ID %s)" % (name, filename, line, _exc_id)
+
+        return cls(
+            probe_id=_exc_id,
+            version=0,
+            tags={},
+            source_file=filename,
+            line=line,
+            template=message,
+            segments=[LiteralTemplateSegment(message)],
+            take_snapshot=True,
+            limits=DEFAULT_CAPTURE_LIMITS,
+            condition=None,
+            condition_error_rate=0.0,
+            rate=float("inf"),
+        )
+
+
+@attr.s
+class SpanExceptionSnapshot(Snapshot):
+    exc_id = attr.ib(type=t.Optional[uuid.UUID], default=None)
+
+    @property
+    def data(self):
+        # type: () -> t.Dict[str, t.Any]
+        data = super(SpanExceptionSnapshot, self).data
+
+        data.update({"exception-id": str(self.exc_id)})
+
+        return data
+
+
+def can_capture(span):
+    # type: (Span) -> bool
+    # We determine if we should capture the exception information from the span
+    # by looking at its local root. If we have budget to capture, we mark the
+    # root as "info captured" and return True. If we don't have budget, we mark
+    # the root as "info not captured" and return False. If the root is already
+    # marked, we return the mark.
+    root = span._local_root
```

Comment:
We are using internal span details here, which is not ideal. Perhaps tracing should expose this via an internal API?